### PR TITLE
JP-3708, JP-3771, JP-3791

### PR DESCRIPTION
**Pull request by kmacdonald-stsci**
Resolves [JP-3708](https://jira.stsci.edu/browse/JP-3708)
Resolves [JP-3771](https://jira.stsci.edu/browse/JP-3771)
Resolves [JP-3791](https://jira.stsci.edu/browse/JP-3791)

JP-3708: Add the CRMAG parameter to the optional results product in the C-extension.  This wasn't implemented with the C-extension, nor was it properly handled for multiprocessing.

JP-3771: Corrected bugs exposed by adding multiprocessing regression tests to the JWST regression testing suite.  The primary bug found was incorrect dimension sizes for internal arrays used for the optional results product.  The internal arrays should have dimensions of `(nints, nrows, ncols)`, but they had dimensions `(ngroups, nrows, ncols)`.  These incorrect dimensions caused memory corruption causing processes to freeze or segfault.

JP-3791: Corrected the CHARGELOSS recomputation of the read noise variance.  For multiprocessing the DQ flags were not being properly sliced and passed to the C-extension for recomputation.

## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] run regression tests with this branch installed (`"git+https://github.com/<fork>/stcal@<branch>"`)
    - [ ] [`jwst` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) 
    - [ ] [`romancal` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.apichange.rst``: change to public API
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
</details>

### CODE REVIEW

File: src/stcal/ramp_fitting/src/slope_fitter.c [hunk 1]
```diff
@@ -784,10 +850,55 @@ print_pid_info(long long prev, int line, char * label) {
...
+ * Redirection from the command line behaves unexpectedly when multiple processes print
+ * to the terminal and copy and pasting from the terminal is lacking due to the amount
+ * of data that possibly needs to be copied.
+ * 
+ * Each process has its own log and each log has a timestamp, so if you have successive
+ * runs, each can be separated by time, as well as process ID.
+ */
+void
+set_up_logger() {
+    const char * log_dir = "/Users/kmacdonald/code/stcal/logs";
```
Comment (stscirij) [inline src/stcal/ramp_fitting/src/slope_fitter.c:875]: Maybe get this location from an environment variable or some other way?

Comment (kmacdonald-stsci) [reply src/stcal/ramp_fitting/src/slope_fitter.c:875]: It's a debugging function that isn't used anywhere.  I left it there so I can use it in the future for debugging.  I'll simply remove the path.

**Comment by kmacdonald-stsci** [comment]
> Looking good!
> 
> Testing on the sample data that failed in #304, I no longer see the processing hang with the optional output + multiprocessing. Optional output looks the same with and without multiprocessing now, and also shows only minor differences from output with the main branch + algorithm = 'OLS'.
> 
> Did you run regression tests for jwst and romancal? If so, can you please link them here?

Here is a link to the regression tests I ran on these PRs:

https://github.com/spacetelescope/RegressionTests/actions/runs/11820513473

These changes don't effect RomanCal, but I will ask a RomanCal maintainer to run regression tests against these changes.

**Review by melanieclarke — APPROVED** [review]
Regression tests for jwst look good.  Docs are building cleanly now -- thanks for the change notes fix!

I have a couple more minor copy editing suggestions for the change note, but I will approve now, from my perspective.  We should still wait to make sure romancal tests are clean.

File: changes/318.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+This PR fixes JP-3791, JP-3771, and JP-3708.  For `ramp_fitting`, the
+`CRMAG` element was not originally implementd in the C-extension for
+ramp fitting.  It is now implemented.  The bug the read noise recalculation
+for CHARGELOSS when using the multiprocessing option has been fixed.  Further,
```
[the remark this answers is not in the document]
Comment (kmacdonald-stsci) [reply changes/318.bugfix.rst]: Why "other than NIRISS"?  This CHARGELOSS recalculation is done in NIRISS.  I chose `test_niriss_image.py` test because the data used for the `rate` tests contain CHARGELOSS flags.

[the remark this answers is not in the document]
Comment (melanieclarke) [reply changes/318.bugfix.rst]: Am I misremembering?  I thought the issue from the user's perspective was that multiprocessing + any instrument other than NIRISS threw the `pr->orig_gdq is NULL` error, because chargeloss corrections were not initialized properly.

[the remark this answers is not in the document]
Comment (kmacdonald-stsci) [reply changes/318.bugfix.rst]: The error thrown during multiprocessing because `pr->orig_gdq` occurred because the CHARGELOSS flag was set, but the data got sliced the `orig_gdq` didn't get sliced and was always always passed to the C-extension as `NULL`.  Below is the fix for that.

https://github.com/spacetelescope/stcal/blob/ebf46dd2023c1b2c23ee97badc93c5928786844e/src/stcal/ramp_fitting/ols_fit.py#L581

[the remark this answers is not in the document]
Comment (melanieclarke) [reply changes/318.bugfix.rst]: Okay, you can just leave that suggestion out then!

[the remark this answers is not in the document]
Comment (kmacdonald-stsci) [reply changes/318.bugfix.rst]: You want me to leave out that this PR fixes a bug found in multiprocessing?

File: src/stcal/ramp_fitting/ols_fit.py
Comment (jemorrison) [inline src/stcal/ramp_fitting/ols_fit.py:1]: Minor comment: 'algorithm" that is in the list of parameters is not in the list of options to old_ramp_fit_multi

File: src/stcal/ramp_fitting/src/slope_fitter.c [hunk 3]
```diff
@@ -753,7 +818,8 @@ is_pix_in_list(struct pixel_ramp * pr)
     cols[0] = 173;
```
Comment (jemorrison) [inline src/stcal/ramp_fitting/src/slope_fitter.c:818]: I take it this hard coded values is just for testing ? Should it be removed now or does it not matter ?

**Review by jemorrison — APPROVED** [review]
I left a few comments - all minor some just for my benefit. 
It all seems reasonable to me. There are a lot of changes in the PR.

File: src/stcal/ramp_fitting/src/slope_fitter.c (hunk 3 above)
Comment (kmacdonald-stsci) [reply src/stcal/ramp_fitting/src/slope_fitter.c:818]: Yes.  The function `is_pix_in_list` is a debugging function.  I use it to validate various properties of a specific pixel being processed.  I normally keep debugging functions in the code, even though they are not called when the code is cleaned up for a PR.  I marked this functions as a debugging function.

File: src/stcal/ramp_fitting/ols_fit.py [hunk 4]
```diff
@@ -570,6 +575,14 @@ def slice_ramp_data(ramp_data, start_row, nrows):
     ramp_data_slice.flags_saturated = ramp_data.flags_saturated
     ramp_data_slice.flags_no_gain_val = ramp_data.flags_no_gain_val
     ramp_data_slice.flags_unreliable_slope = ramp_data.flags_unreliable_slope
+    ramp_data_slice.flags_chargeloss = ramp_data.flags_chargeloss
+
+    # For possible CHARGELOSS flagging.
+    if ramp_data.orig_gdq is not None:
+        ogdq = ramp_data.orig_gdq[:, :, start_row : start_row + nrows, :].copy()
```
[the remark this answers is not in the document]
Comment (kmacdonald-stsci) [reply src/stcal/ramp_fitting/ols_fit.py:578]: The `orig_gdq` is a copy of the group DQ array that comes out of jump, so it's the same shape as a `RampModel` array if `(nints, ngroups, nrows, ncols)`.  Because ramp fitting modifies this DQ array (for example, the CHARGELOSS (and associated DO_NOT_USE) flag is removed during ramp fitting, the original needs to be copied to recompute the read noise variance for ramps with flagged as CHARGELOSS.

**Review by tapastro — APPROVED** [review]
